### ISSUE TITLE
[0.21 or 0.22] Migrate all console calls from printf style to fmt::format style

### ISSUE THREAD

**Issue by abdullahtahiriyo** [FreeCAD/FreeCAD-translations#226]
Why?
https://www.zverovich.net/slides/2017-cppcon.pdf

Scope:
FreeCAD-wide

Impact:
Every single call to Console().Log, Console().Warning, Console().XXXX

Difficulty:
Low (basically migrate the format specification from the %s %d to {})

**Comment by abdullahtahiriyo** [comment]
@chennes 
@wwmayer 

Eventually we should also look into this.

This relates to:
https://github.com/FreeCAD/FreeCAD-translations/issues/223
https://forum.freecad.org/viewtopic.php?p=683602#p683602

Basically, we are trying to solve the issue of the dynamic strings by storing the format specification and the arguments for exceptions. If we introduce format specification and arguments for exceptions, the future appears to be fmt::format. Then, we should avoid having a plethora of formats lying around.

N.B.: The presentation (pg. 69) talks about using clang-tidy for automatic transform code. Not sure if this facility can be used for the console part.

**Comment by abdullahtahiriyo** [comment]
It is quite a daunting task.

Today I wanted to estimate the amount of changes, as well as try to identify issues.

I run this in the tree:
```find ./ -type f -exec sed -i -e '/Console/s/%./\{\}/g' {} \;```

and changed Console to process the formatter string as format.

Of course this does not cover all possible patterns of %.

After that, there are still some 67 appearances of "%s" after that massive change (```grep -r -i "%s" *```). 14 of "%d. 12 of "%p". 1 of "%f".

I have looked them one by one. 

Most were legit, such as use of printf, fprintf, snprintf or PErr_Format, e.g. in ConsoleObserver.cpp or MemDebug.cpp, or even 3rd party libraries. 

Some are due to the Console statement being across several lines, which the simplistic shell one liner does not handle.

There may be other changes that are not so straightforward (separate formatter string in a variable, or macros indirectly calling the function).

I provide here as a draft commit to assess the impact and how we could approach this... and when...

https://github.com/FreeCAD/FreeCAD/pull/9688

**Comment by berniev** [comment]
Have you looked at `fmt::printf` ? (I havn't used it myself)

**Comment by abdullahtahiriyo** [comment]
@berniev 

Yes. It is what we have now in master (I introduced it as part of the Console redesign a couple of months ago).

Part of the idea is to move away from the printf syntax, towards the format syntax.

**Comment by kaktusus** [comment]
Is this ticket can be closed?

**Comment by chennes** [comment]
As written, yes: it's an attempt at solving the non-translation of complex exception strings, but unfortunately in C++20 the proposed mechanism does not work. See https://github.com/FreeCAD/FreeCAD/pull/9688#issuecomment-2692560870
